src/NServiceBus.Core/Transports/Msmq/ReceiveWithTransactionScope.cs

```diff
@@ -17,45 +18,81 @@ public ReceiveWithTransactionScope(TransactionOptions transactionOptions)
-                    Logger.Error(error, ex);
+                    try
+                    {
+                        headers = MsmqUtilities.ExtractHeaders(message);
+                    }
+                    catch (Exception ex)
+                    {
+                        var error = $"Message '{message.Id}' is corrupt and will be moved to '{errorQueue.QueueName}'";
+                        Logger.Error(error, ex);
 
-                    errorQueue.Send(message, MessageQueueTransactionType.Automatic);
+                        errorQueue.Send(message, MessageQueueTransactionType.Automatic);
 
-                    scope.Complete();
-                    return;
-                }
+                        scope.Complete();
+                        return;
+                    }
 
-                using (var bodyStream = message.BodyStream)
-                {
-                    var ambientTransaction = new TransportTransaction();
-                    ambientTransaction.Set(Transaction.Current);
-                    var pushContext = new PushContext(message.Id, headers, bodyStream, ambientTransaction, cancellationTokenSource, new ContextBag());
+                    using (var bodyStream = message.BodyStream)
+                    {
+                        var ambientTransaction = new TransportTransaction();
+                        ambientTransaction.Set(Transaction.Current);
+                        var pushContext = new PushContext(message.Id, headers, bodyStream, ambientTransaction, cancellationTokenSource, new ContextBag());
 
-                    await onMessage(pushContext).ConfigureAwait(false);
-                }
+                        Exception ex;
+                        int attemptNumber;
+                        if (!HasFailedBefore(message.Id, out ex, out attemptNumber))
+                        {
+                            await onMessage(pushContext).ConfigureAwait(false);
+                        }
+                        else
+                        {
+                            var shouldRetryImmediately = await onError(new ErrorContext(ex, attemptNumber, ambientTransaction, message.Id, headers, bodyStream)).ConfigureAwait(false);
+                            if (shouldRetryImmediately)
+                            {
+                                await onMessage(pushContext).ConfigureAwait(false);
+                            }
+                        }
+                    }
 
-                if (!cancellationTokenSource.Token.IsCancellationRequested)
+                    if (!cancellationTokenSource.Token.IsCancellationRequested)
+                    {
+                        scope.Complete();
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                if (message == null)
                 {
-                    scope.Complete();
+                    throw;
                 }
+                RecordException(message.Id, ex);
```

Comment:
This will cause infinite-loop when we start throwing (for some reason) from `onError`